src/onlineservice/sopcast_content_handler.cc

```diff
@@ -39,193 +39,187 @@
 #include "online_service.h"
 #include "util/tools.h"
 
-using namespace zmm;
-using namespace mxml;
-
 SopCastContentHandler::SopCastContentHandler(std::shared_ptr<ConfigManager> config, std::shared_ptr<Storage> storage)
     : config(config)
     , storage(storage)
 {
 }
 
-bool SopCastContentHandler::setServiceContent(zmm::Ref<mxml::Element> service)
+void SopCastContentHandler::setServiceContent(std::unique_ptr<pugi::xml_document>& service)
 {
-    std::string temp;
+    service_xml = std::move(service);
+    auto root = service_xml->document_element();
 
-    if (service->getName() != "channels")
+    if (std::string(root.name()) != "channels")
         throw std::runtime_error("Invalid XML for SopCast service received");
 
-    channels = service;
-
-    group_count = channels->childCount();
-    if (group_count < 1)
-        return false;
-
-    current_group_node_index = 0;
-    current_channel_index = 0;
-    channel_count = 0;
-    current_group = nullptr;
-
-    return true;
+    group_it = root.begin();
+    if (group_it != root.end())
+        channel_it = (*group_it).begin();
 }
 
 std::shared_ptr<CdsObject> SopCastContentHandler::getNextObject()
 {
-#define DATE_BUF_LEN 12
-    std::string temp;
-    struct timespec ts;
-
-    while (current_group_node_index < group_count) {
-        if (current_group == nullptr) {
-            Ref<Node> n = channels->getChild(current_group_node_index);
-            current_group_node_index++;
+    auto root = service_xml->document_element();
+    bool skipGroup = true;
```

Comment:
this is not correct for the first call of getNextObject, i need to fix this first.
like:

skipGroup = false
while (...)
  if (skipGroup || ...) {
     ...
     skipGoup = false; // we are on the next group
     continue;
  }
  skipGroup = true;
  ...